### PR TITLE
add breakout training example

### PR DESCRIPTION
**Pull request by hnyu**
1. 10 hours to reach 300 on breakout (actions are sampled). Wall clock time needs to be improved a lot
2. add advantage clipping and reward shaping options
3.  use gin files in root_dir if none is provided

### CODE REVIEW

File: alf/algorithms/actor_critic_algorithm.py [hunk 1]
```diff
@@ -71,7 +73,10 @@ def __init__(self,
           extrinsic_reward_coef: Coefficient for extrinsic reward
           loss (None|ActorCriticLoss): an object for calculating loss. If None,
             a default ActorCriticLoss will be used.
-          optimizer (tf.optimizers.Optimizer): The optimizer for training.
+          optimizer (tf.optimizers.Optimizer): The optimizer for training
+          gradient_clipping (float): positive threshold for clipping gradient norms
```
Comment (emailweixu) [inline alf/algorithms/actor_critic_algorithm.py]: If not None,

File: alf/algorithms/on_policy_algorithm.py [hunk 2]
```diff
@@ -265,10 +270,18 @@ def train_complete(
...
             loss_info (LossInfo): loss information
             grads_and_vars (list[tuple]): list of gradient and variable tuples
-            
+
         """
         valid_masks = tf.cast(
             tf.not_equal(training_info.step_type, StepType.LAST), tf.float32)
+
+        # reward shaping
+        if self._reward_shaping_fn is not None:
```
Comment (emailweixu) [inline alf/algorithms/on_policy_algorithm.py]: summarize the original reward

File: alf/examples/ac_breakout.gin [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+# "v0 vs v4: v0 has repeat_action_probability of 0.25
+#  (meaning 25% of the time the previous action will be used instead of the new action),
+#   while v4 has 0 (always follow your issued action)
+# Because we already implements frame_skip in AtariPreprocessing, we should always
+# use 'NoFrameSkip' Atari environments from OpenAI gym
+create_environment.env_name='BreakoutNoFrameskip-v4'
+create_environment.env_load_fn=@suite_atari.load
+create_environment.num_parallel_environments=512
+
+# Do not use this frame_skip! Potential issue with _pool_and_resize()
```
Comment (emailweixu) [inline alf/examples/ac_breakout.gin]: I don't understand this comment. If it should not be used, why it is here.

File: alf/examples/actor_critic.py [hunk 4]
```diff
@@ -222,12 +221,17 @@ def copy_gin_configs(root_dir, gin_files):
...
+
+    if FLAGS.gin_file is None:
+        gin_file = glob.glob(FLAGS.root_dir + "/*.gin")
+        assert gin_file, "No gin files are found! Please provide"
+
     if not FLAGS.play:
         copy_gin_configs(FLAGS.root_dir, FLAGS.gin_file)
-        # TODO: use the gin files under FLAGS.root_dir if FLAGS.gin_file
-        # is not specified
+        gin_file = FLAGS.gin_file
```
Comment (emailweixu) [inline alf/examples/actor_critic.py]: if FLAGS.gin_file is None, gin_file will become None

File: alf/trainers/on_policy_trainer.py [hunk 5]
```diff
@@ -88,8 +88,8 @@ def train_():
             driver.run = tf.function(driver.run)
 
         env.reset()
-        time_step = driver.get_initial_time_step()
-        policy_state = driver.get_initial_state()
+        time_step = None
+        policy_state = None
```
Comment (emailweixu) [inline alf/trainers/on_policy_trainer.py]: I assign the initial value for time_step and policy_state for the reason of keeping the calling signature to driver.run() same. tf.function will recompile the graph based on the calling signature.

File: alf/utils/common.py [hunk 6]
```diff
@@ -259,3 +259,26 @@ def _transform_image(obs):
...
+
+    Can be used as a reward shaping function passed to an algorithm
+    (e.g. ActorCriticAlgorithm).
+    """
+    assert minmax[0] <= minmax[1], "range error"
+    return tf.clip_by_value(r, minmax[0], minmax[1])
+
+
+@gin.configurable
+def reward_scaling(r, scale=1):
```
Comment (emailweixu) [inline alf/utils/common.py]: simply scaling the reward won't have any meaningful effect because the gradient is proportionally scaled.

File: alf/algorithms/actor_critic_algorithm.py (hunk 1 above)
Comment (hnyu) [reply alf/algorithms/actor_critic_algorithm.py]: added

File: alf/algorithms/on_policy_algorithm.py (hunk 2 above)
Comment (hnyu) [reply alf/algorithms/on_policy_algorithm.py]: added

File: alf/examples/ac_breakout.gin (hunk 3 above)
Comment (hnyu) [reply alf/examples/ac_breakout.gin]: updated

File: alf/examples/actor_critic.py (hunk 4 above)
Comment (hnyu) [reply alf/examples/actor_critic.py]: fixed

File: alf/trainers/on_policy_trainer.py (hunk 5 above)
Comment (hnyu) [reply alf/trainers/on_policy_trainer.py]: reverted

File: alf/utils/common.py (hunk 6 above)
Comment (hnyu) [reply alf/utils/common.py]: Depending on the init values of the value functions, the advantages or td errors are not linearly correlated to the reward

File: alf/algorithms/actor_critic_algorithm.py [hunk 7]
```diff
@@ -170,8 +176,6 @@ def calc_loss(self, training_info, final_time_step, final_policy_step):
         ac_loss = self._loss(training_info, training_info.info.value,
                              final_time_step, final_value)
 
-        self.add_reward_summary("reward", training_info.reward)
-
```
Comment (emailweixu) [inline alf/algorithms/actor_critic_algorithm.py]: We should still keep this one, this is reward for training.
I am also thinking that we probably need to document the meaning of `reward` clearly, otherwise it might be used incorrectly, Or perhaps we should use different name to differentiate different reward: 'environment_reward' and 'reward`

Comment (hnyu) [reply alf/algorithms/actor_critic_algorithm.py]: It's not gone. It was moved to the parent class on_policy_algorithm.py with the appropriate name
